src/main/java/org/killbill/billing/client/KillBillHttpClient.java

```diff
@@ -391,87 +391,121 @@ public Response doOptions(final String uri, final Multimap<String, String> optio
 
     private String getUniqueValue(final Multimap<String, String> options, final String key) {
         final Collection<String> values = options.get(key);
-        if (values == null || values.size() == 0) {
+        if (values == null || values.isEmpty()) {
             return null;
         } else {
             Preconditions.checkState(values.size() == 1, "You can only specify a unique value for " + key);
             return values.iterator().next();
         }
     }
 
-    private <T> T executeAndWait(final BoundRequestBuilder builder, final int timeoutSec, final Class<T> clazz) throws KillBillClientException {
-        Response response;
-        final ListenableFuture<Response> futureStatus;
+    private Response request(final BoundRequestBuilder builder, final int timeoutSec) throws KillBillClientException {
         try {
-            futureStatus = builder.execute(new AsyncCompletionHandler<Response>() {
+            final ListenableFuture<Response> futureStatus = builder.execute(new AsyncCompletionHandler<Response>() {
                 @Override
                 public Response onCompleted(final Response response) throws Exception {
                     return response;
                 }
             });
-            response = futureStatus.get(timeoutSec, TimeUnit.SECONDS);
+            return futureStatus.get(timeoutSec, TimeUnit.SECONDS);
         } catch (final InterruptedException e) {
             throw new KillBillClientException(e);
         } catch (final ExecutionException e) {
             throw new KillBillClientException(e);
         } catch (final TimeoutException e) {
             throw new KillBillClientException(e);
         }
+    }
 
-        if (response != null && response.getStatusCode() == 401) {
-            throw new KillBillClientException(new IllegalArgumentException("Unauthorized - did you configure your RBAC and/or tenant credentials?"),
-                                              response);
-        } else if (response != null && (response.getStatusCode() == 404 || response.getStatusCode() == 204)) {
-            // Return empty list for KillBillObjects instead of null for convenience
-            if (Iterable.class.isAssignableFrom(clazz)) {
-                for (final Constructor constructor : clazz.getConstructors()) {
-                    if (constructor.getParameterTypes().length == 0) {
-                        try {
-                            return (T) constructor.newInstance();
-                        } catch (InstantiationException e) {
-                            return null;
-                        } catch (IllegalAccessException e) {
-                            return null;
-                        } catch (InvocationTargetException e) {
-                            return null;
-                        }
-                    }
-                }
-                return null;
+    private void throwExceptionOnResponseError(final Response response) throws KillBillClientException {
```

Comment:
I find the code confusing (maybe it was already confusing before). From what i understand at this point you already took care of the `404` and `204` status. At this point we now treat `401` and then default for all other errors > `400`.
Also that method sometimes throw and sometimes does not not which means we need to still return something after for the cases where it does not throw.

Would it make sense to have one switch statement in **one place** that handles each http status?